---
title: "Integration_Test_Kidney"
output: html_document
date: '2022-11-30'

---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(devtools)
library(rprojroot)
load_all()

root <- rprojroot::find_package_root_file()

knitr::opts_chunk$set(fig.width=12, fig.height=8, out.width = '100%') 
knitr::opts_knit$set(root.dir = root)


```

## R Markdown

This runs the DSPworkflow package to completion using the Human Kidney Dataset:

## 1. Study Design:

```{r Study Design, echo=FALSE}
  
  # Set paths for downloading dcc files
  downloads.path <- test_path("fixtures/Mouse_Thymus/downloaded/")
  tar.file.name <- "thymus_dccs.tar.gz"
  full.tar.path <- paste0(downloads.path,tar.file.name)
  
  # Check if dcc files were previously downloaded
  if (!file.exists(full.tar.path)) {
    
    # Download dcc files and place in data folder
    data.url <- "http://hpc.nih.gov/~CCBR/DSPWorkflow/thymus_dccs.tar.gz"
    download.file(data.url, full.tar.path)
    untar(full.tar.path, exdir = downloads.path)
  }
  
  dcc.files <- dir(
    file.path(
      downloads.path,
      "dccs"
    ),
    pattern = ".dcc$",
    full.names = TRUE,
    recursive = TRUE
  )
    
  pkc.files <-
    test_path("fixtures/Mouse_Thymus/Mm_R_NGS_WTA_v1.0.pkc")
  pheno.data.file <-
    test_path("fixtures/Mouse_Thymus/Thymus_Annotation_updated_3.xlsx")
  
  sdesign.list <- studyDesign(dcc.files = dcc.files, 
                                pkc.files = pkc.files,
                                pheno.data.file = pheno.data.file,
                                pheno.data.sheet = "Annotation",
                                pheno.data.dcc.col.name = "Sample_ID",
                                protocol.data.col.names = c("aoi", "roi"),
                                experiment.data.col.names = c("panel"))
  
  # For creating fixture RDS
  create.rds <- FALSE
  if(create.rds) {
    study.design.human.kidney <- sdesign.list$object
    saveRDS(study.design.human.kidney, file = "tests/testthat/fixtures/Mouse_Thymus/studyDesignMouseThymus.RDS")
  }
  
  print(sdesign.list$sankey.plot)
  print("Created GeoMx Object\n\n")
  pData(sdesign.list$object)[,c("slide name","class","segment")]  
  
```

## 2. QC Preprocessing:

```{r QC Preprocessing, echo=FALSE}

qc.output <-  qcProc(object = sdesign.list$dsp.obj,
                        minSegmentReads = 1000, # Default:1000
                        percentTrimmed = 65,  # Default:80  
                        percentStitched = 65, # Default:80 
                        percentAligned = 65, #Default:80
                        percentSaturation = 50, #Default:50
                        minNegativeCount = 10, #Default:10   
                        maxNTCCount = 1000, #Default:1000     
                        minNuclei = 200, #Default:200         
                        minArea = 16000) #Default:16000
    print(QCoutput$segments.qc)
    
qc.output <-  qcProc(object = sdesign.list$object,
                        min.segment.reads = 1000, 
                        percent.trimmed = 80,    
                        percent.stitched = 80,   
                        percent.aligned = 75,    
                        percent.saturation = 50, 
                        min.negative.count = 1,   
                        max.ntc.count = 9000,     
                        min.nuclei = 20,         
                        min.area = 1000,
                        print.plots = TRUE)
    print(qc.output$segments.qc)
    
  create.rds <- FALSE
  if(create.rds) {
    qc.human.kidney <- qc.output$object
    saveRDS(qc.human.kidney, file = "tests/testthat/fixtures/Human_Kidney/qcHumanKidney.RDS")

```

## 3. Filtering:

```{r Filtering, echo=FALSE}
    
    PKFil <- annotation(QCoutput$dsp.target)
    target_demoDataFil <- QCoutput$dsp.target
    
    genes <- c("Plb1", "Ccr7", "Oas2", "Oas1a", "Oas1b", "Rhbdl2", "Dlst", "Naa15", "Rab11a", "Desi1", "Tfdp1", "Foxn1")
    
    FiltOutput <- filtering(Data = target_demoDataFil, 
                            PKCS = PKFil, 
                            LOQcutoff = 2, 
                            LOQmin = 2, 
                            CutSegment = .05, 
                            GOI = genes)
    
    print(FiltOutput$`Stacked Bar Plot`)
    print(FiltOutput$`Table of Cuts`)
    print(FiltOutput$`Sankey Plot`)
    print(FiltOutput$`Genes Deccted Plot`)
    
    #filteringMouseThymus <- FiltOutput$`target_demoData Dataset`
    
    #saveRDS(filteringMouseThymus, file = "filteringMouseThymus.RDS")

```


## 4. Normalization:
  
```{r Normalization, echo=FALSE}
  
    target_demoDataNorm <- FiltOutput$`target_demoData Dataset`
    
    NormOut.list <- GeoMxNorm(target_demoDataNorm, "quant")
    
    print(NormOut.list$plot)
    print(NormOut.list$Boxplot)
    
    #normalizationMouseThymus <- NormOut.list$`Normalized Dataframe`
    
    #saveRDS(normalizationMouseThymus, file = "normalizationMouseThymus.RDS")
    
```


## 5. Unsupervised Analysis:

```{r Unsupervised Analysis, echo=FALSE}

    #Test Unsupervised Analysis:
    target.Data <- NormOut.list$`Normalized Dataframe`
    UnSupervisedoutput <- DimReduct(object = target.Data,
                        point.size = 2,
                        point.alpha = 1,
                        color.variable1 = "region",
                        shape.variable = "class"
    )
    
    print(UnSupervisedoutput$plot.list$PCA)
    print(UnSupervisedoutput$plot.list$tSNE)
    print(UnSupervisedoutput$plot.list$UMAP)
    
    #unsupervisedAnalysisMouseThymus <- UnSupervisedoutput$dsp.object
    
    #saveRDS(unsupervisedAnalysisMouseThymus, file = "unsupervisedAnalysisMouseThymus.RDS")

```


## 6. Clustering high CV Genes and Heatmap:

 
```{r Clustering high CV Genes, echo=FALSE}

    target.Data <- UnSupervisedoutput$dsp.object  
    
    dsp.list <- HeatMap(target.data = target.Data, 
                        ngenes = 200, 
                        image.width = 3600, 
                        image.height = 1800, 
                        image.resolution = 300, 
                        image.filename = "heatmap.clust.highCVgenes.png",
                        scale.by.row.or.col = "row", 
                        show.rownames = FALSE, 
                        show.colnames = FALSE, 
                        clustering.method = "average", 
                        cluster.rows = TRUE, 
                        cluster.cols = TRUE,
                        clustering.distance.rows = "correlation", 
                        clustering.distance.cols = "correlation", 
                        annotation.row = NA, 
                        annotation.col = c("class", "segment", "region"),
                        breaks.by.values = seq(-3, 3, 0.05), 
                        heatmap.color = colorRampPalette(c("blue", "white", "red"))(120), 
                        norm.method = "quant")


    print(dsp.list$plot)
    
  
```


## 7. Differential Expression Analysis:


```{r Differential Expression Analysis,  echo=FALSE}

    data <- UnSupervisedoutput$dsp.object 
    goi <- c("Plb1", "Ccr7", "Oas2", "Oas1a", "Oas1b", "Rhbdl2", "Dlst", "Naa15", "Rab11a", "Desi1", "Tfdp1", "Foxn1")
    
    data <- data[goi,]
    Gene <- Subset <- NULL
    
    #First analysis:
    reslist.1 <- DiffExpr(object = data, 
                          analysisType = "Within Groups", 
                          regionCol = "region", 
                          regions = c("Cortical", "Medullar"), 
                          groupCol = "class", 
                          groups = c("Thymus"), 
                          slideCol = "slide name")
    grid.draw(reslist.1$sample_table)
    grid.newpage()
    grid.draw(reslist.1$summary_table)
    
    lfc_col1 <- colnames(reslist.1$result)[grepl("logFC",colnames(reslist.1$result))]
    pval_col1 <- colnames(reslist.1$result)[grepl("_pval",colnames(reslist.1$result))]
    
    lfc.1 <- reslist.1$result %>% 
              dplyr::filter(Gene == "Ccr7" & Subset == "Thymus") %>% 
              select(all_of(lfc_col1)) %>% 
              as.numeric()
    pval.1 <- reslist.1$result %>% 
              dplyr::filter(Gene == "Ccr7" & Subset == "Thymus") %>% 
              select(all_of(pval_col1)) %>% 
              as.numeric()
    
    cat(paste0("\n\nvalue of Ccr7 Fold Change is:", lfc.1))
    cat("expected value is -1.645")
    
    cat(paste0("\nvalue of Ccr7 pval is:",pval.1))
    cat("expected value is 0.0274")
    
    #Second analysis:
    reslist.2 <- DiffExpr(object = data,
                          analysisType = "Between Groups", 
                          regionCol = "segment", 
                          regions = c("PanCK"), 
                          groupCol = "region", 
                          groups = c("Tumor", "Medullar"), 
                          slideCol = "slide name")
    grid.draw(reslist.2$sample_table)
    grid.newpage()
    grid.draw(reslist.2$summary_table)
    
    lfc_col2 <- colnames(reslist.2$result)[grepl("logFC",colnames(reslist.2$result))]
    pval_col2 <- colnames(reslist.2$result)[grepl("_pval",colnames(reslist.2$result))]
    
    lfc.2 <- reslist.2$result %>% 
              dplyr::filter(Gene == "Ccr7" & Subset == "PanCK") %>% 
              select(all_of(lfc_col2)) %>% 
              as.numeric()
    pval.2 <- reslist.2$result %>% 
              dplyr::filter(Gene == "Ccr7" & Subset == "PanCK") %>% 
              select(all_of(pval_col2)) %>% 
              as.numeric()
    
    cat(paste0("\n\nvalue of Ccr7 Fold Change is: ", lfc.2))
    cat("expected value is -1.89")
    
    cat(paste0("\nvalue of Ccr7 pval is: ",pval.2))
    cat("expected value is 4.97e-12")
    
    print(reslist.2$results)

```
## 8. Volcano Plot

#This part is run on NIDAP.

## 9. Violin Plot

```{r Violin Plot, echo=FALSE}

    test_data <- UnSupervisedoutput$dsp.object
    
    test_genes <- head(rownames(test_data@assayData$q_norm),5)
    
    #test_genes <- c("Plb1", "Ccr7", "Oas2", "Oas1a", "Oas1b", "Rhbdl2", "Dlst", "Naa15", "Rab11a", "Desi1", "Tfdp1", "Foxn1")
    
    #test_genes <- c("Desi1", "Tfdp1", "Foxn1")

    #test_genes <- c("Plb1", "Desi1")
    
    violin_plot_test <- violin_function(data = test_data, 
                                        genes = test_genes,
                                        group = "region")
    grid.arrange(violin_plot_test)
    
    print("Violin Plot Done")

```

## 10. Spatial Deconvolution:

```{r Spatial Deconvolution, echo=FALSE}

refdir <- "/rstudio-files/ccr-dceg-data/data/CellProfileLibrary/Mouse/Adult/"
    load(paste0(refdir,"Thymus_MCA.RData"))
    ref_mtx_test <- profile_matrix
    annot_1 <- data.frame(CellID = colnames(ref_mtx_test),LabeledCellType = colnames(ref_mtx_test))
    
    target_demoData <- UnSupervisedoutput$dsp.object
    dsp_qnorm_test = target_demoData@assayData$q_norm
    dsp_negnorm_test = target_demoData@assayData$q_norm
    
    #To derive neg probes: 
    target_demoData@featureData@data$TargetName[target_demoData@featureData@data$Negative == TRUE] 
    
    dsp_negnorm_test = derive_GeoMx_background(norm = target_demoData@assayData$q_norm,
                                 probepool = fData(target_demoData)$Module,
                                 negnames = c("NegProbe-WTX"))
    
    spatial_output <- spatial_deconvolution(dsp_qnorm = dsp_qnorm_test, 
                                  dsp_negnorm = dsp_negnorm_test, 
                                  ref_mtx = ref_mtx_test, ref_annot = annot_1)
    
    print(spatial_output$figures)

    print("Spatial Deconvolution Done")

```
